extensions/notebook/src/book/bookTreeView.ts

```diff
@@ -0,0 +1,119 @@
+/*---------------------------------------------------------------------------------------------
+ *  Copyright (c) Microsoft Corporation. All rights reserved.
+ *  Licensed under the Source EULA. See License.txt in the project root for license information.
+ *--------------------------------------------------------------------------------------------*/
+
+import * as vscode from 'vscode';
+import * as path from 'path';
+import * as fs from 'fs';
+import * as yaml from 'js-yaml';
+import { Book } from './bookTreeItem';
+import * as nls from 'vscode-nls';
+const localize = nls.loadMessageBundle();
+
+
+export class BookTreeViewProvider implements vscode.TreeDataProvider<Book> {
+
+	private _onDidChangeTreeData: vscode.EventEmitter<Book | undefined> = new vscode.EventEmitter<Book | undefined>();
+	readonly onDidChangeTreeData: vscode.Event<Book | undefined> = this._onDidChangeTreeData.event;
+	private tableOfContentsPath: string[];
+
+	constructor(private workspaceRoot: string) {
+		if (workspaceRoot !== '') {
+			this.tableOfContentsPath = this.getFiles(this.workspaceRoot, []);
+			if (this.tableOfContentsPath === undefined || this.tableOfContentsPath.length === 0) {
+				vscode.commands.executeCommand('setContext', 'bookOpened', false);
+			} else {
+				vscode.commands.executeCommand('setContext', 'bookOpened', true);
+			}
+		}
+	}
+
+	private getFiles(dir: string, allFiles: string[]): string[] {
+		let files = fs.readdirSync(dir);
+		for (let i in files) {
+			let name = path.join(dir, files[i]);
+			if (fs.statSync(name).isDirectory()) {
+				this.getFiles(name, allFiles);
+			} else if (files[i] === 'toc.yml') {
+				allFiles.push(name);
+			}
```

Comment:
Instead of passing in the array in each why not just start with an empty list and push all new files and return that - then append the return value from all recursive calls of this? Does the same thing and saves having to pass that array around.